### DISCUSSION TITLE
[REQUEST] Enable specifying `requirements.in` & ignore `requirements.txt` for `pip_requirements`

### DISCUSSION THREAD

**Discussion by rahul-theorem** [renovatebot/renovate#17209] [Request Help] [answered]
### What would you like Renovate to be able to do?

We currently use Bazel's `rules_python` to manage python requirements in our monorepo. We use a `requirements.in` file to specify our requirements, which is used to generate a lockfile (`requirements.txt`). This is generated using `pip-compile` under the hood, but the format of the `requirements.txt` is specific to Bazel (see https://github.com/renovatebot/renovate/issues/16283 for a related issue).

When we specify the `pip_compile` manager in `renovate.json`, we encounter the following error:
```
        Building wheels for collected packages: numpy
          Building wheel for numpy (setup.py): started
          Building wheel for numpy (setup.py): finished with status 'error'
          error: subprocess-exited-with-error
        
          × python setup.py bdist_wheel did not run successfully.
          │ exit code: 1
          ╰─> [303 lines of output]
              Running from numpy source directory.
              blas_opt_info:
              blas_mkl_info:
              customize UnixCCompiler
                libraries mkl_rt not found in ['/opt/buildpack/tools/python/3.10.6/lib', '/usr/local/lib', '/usr/lib64', '/usr/lib']
                NOT AVAILABLE
```

Note that Renovate is able to upgrade our `requirements.in` correctly, but fails to produce the `requirements.txt`. We're currently working around this with a custom GitHub action that re runs `bazel run //:requirements.update` on any change to `requirements.in` & pushes a commit to the branch.

In the short-term, we'd like to disable the behavior for renovate to run `pip-compile` so we can avoid this error message being added to each upgrade PR. We've tried changing the manager to `pip_requirements` & specifying `requirements.in` as the pattern, but when using this manager Renovate appears to totally ignore `requirements.in` & instead only upgrades `requirements.txt`. Is there a way we can accomplish this using `pip_requirements`? Failing that, would it be possible to support running `bazel run //:requirements.update` as an optional step for the `pip_compile` manager?

### If you have any ideas on how this should be implemented, please tell us here.

N/A

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by rarkins** [comment]
I can't think ignore reason why your first approach doesn't work unless it's a config error. Can you reproduce it publicly?

**Comment by rahul-theorem** [comment]
Thanks for the reply @rarkins - here's a public repro of the issue: https://github.com/theoremlp/renovate-bug-repro/pull/1

**Comment by rarkins** [comment]
@rahul-theorem I was expecting to see this part:

> We've tried changing the manager to pip_requirements & specifying requirements.in as the pattern, but when using this manager Renovate appears to totally ignore requirements.in & instead only upgrades requirements.txt.

**Comment by markelliot** [comment]
After changing the manager to pip_requirements and setting the fileMatcher pattern [per the docs](https://docs.renovatebot.com/python/#alternative-file-names) to "requirements.in" we're only seeing updates to requirements.txt:
https://github.com/theoremlp/renovate-bug-repro/pull/4

Did we configure this incorrectly?

**Comment by rarkins** [comment, marked as answer]
FYI if you log into app.renovatebot.com then you can see logs for each run.

Here's what I see when I put this through a dry run locally:

```
{"name":"renovate","hostname":"Rhys-MBP.lan","pid":18712,"level":20,"logContext":"63RCX8cl06bvn4RlL9G-y","repository":"theoremlp/renovate-bug-repro","msg":"Using file match: (^|/)([\\w-]*)requirements\\.(txt|pip)$ for manager pip_requirements","time":"2022-08-13T05:38:19.400Z","v":0}
{"name":"renovate","hostname":"Rhys-MBP.lan","pid":18712,"level":20,"logContext":"63RCX8cl06bvn4RlL9G-y","repository":"theoremlp/renovate-bug-repro","msg":"Using file match: requirements.in for manager pip_requirements","time":"2022-08-13T05:38:19.401Z","v":0}
{"name":"renovate","hostname":"Rhys-MBP.lan","pid":18712,"level":20,"logContext":"63RCX8cl06bvn4RlL9G-y","repository":"theoremlp/renovate-bug-repro","msg":"Matched 2 file(s) for manager pip_requirements: requirements.in, requirements.txt","time":"2022-08-13T05:38:19.401Z","v":0}
```

Unfortunately the `fileMatch` is being *merged* instead of overwritten. This is by design, because most of the time people want to add a pattern without needing to repeat all the existing patterns.

You need to add this to your config:

```
  "ignorePaths": ["requirements.txt"]
```

**Comment by markelliot** [comment]
Thanks -- that's fixed things for us. We're a little puzzled why renovate's pip-compile runs won't generate an updated lockfile for us given we're able to run pip-compile via Bazel to get an update, but that's probably a separate issue.
